Source/WebCore/dom/Document.cpp

```diff
@@ -3767,70 +3786,73 @@ void Document::setBaseURLOverride(const URL& url)
 void Document::processBaseElement()
 {
     // Find the first href attribute in a base element and the first target attribute in a base element.
-    const AtomString* href = nullptr;
-    const AtomString* target = nullptr;
+    AtomString href;
+    AtomString target;
     auto baseDescendants = descendantsOfType<HTMLBaseElement>(*this);
     for (auto& base : baseDescendants) {
-        if (!href) {
-            const AtomString& value = base.attributeWithoutSynchronization(hrefAttr);
+        if (href.isNull()) {
+            auto& value = base.attributeWithoutSynchronization(hrefAttr);
             if (!value.isNull()) {
-                href = &value;
-                if (target)
+                href = value;
+                if (!target.isNull())
                     break;
             }
         }
-        if (!target) {
-            const AtomString& value = base.attributeWithoutSynchronization(targetAttr);
+        if (target.isNull()) {
+            auto& value = base.attributeWithoutSynchronization(targetAttr);
             if (!value.isNull()) {
-                target = &value;
-                if (href)
+                target = value;
+                if (!href.isNull())
                     break;
             }
         }
     }
 
     URL baseElementURL;
-    if (href)
-        baseElementURL = completeURL(href->string(), fallbackBaseURL());
+    if (!href.isNull())
+        baseElementURL = completeURL(href, fallbackBaseURL());
     if (m_baseElementURL != baseElementURL) {
         if (!contentSecurityPolicy()->allowBaseURI(baseElementURL))
             m_baseElementURL = { };
         else if (settings().shouldRestrictBaseURLSchemes() && !SecurityPolicy::isBaseURLSchemeAllowed(baseElementURL)) {
             m_baseElementURL = { };
             addConsoleMessage(MessageSource::Security, MessageLevel::Error, "Blocked setting " + baseElementURL.stringCenterEllipsizedToLength() + " as the base URL because it does not have an allowed scheme.");
         } else
-            m_baseElementURL = baseElementURL;
+            m_baseElementURL = WTFMove(baseElementURL);
         updateBaseURL();
     }
 
-    m_baseTarget = target ? *target : nullAtom();
+    m_baseTarget = WTFMove(target);
 }
 
 String Document::userAgent(const URL& url) const
 {
-    return frame() ? frame()->loader().userAgent(url) : String();
+    RefPtr frame = this->frame();
+    return frame ? frame->loader().userAgent(url) : String();
```

Comment:
I think we should be storing loader() in a RefPtr/CheckedPtr since that's the object on which we call a non-trivial function: userAgent. loader() is a trivial function for all intended purposes.

Reply:
FrameLoader is not refcounted. The FrameLoader is owned by the frame so ref'ing the frame (like I did) is what keeps the FrameLoader alive.

Reply:
Can we use a CheckedRef for FrameLoader in that case?